### PR TITLE
`mem_buf` only used with memory BIO

### PR DESCRIPTION
**Pull request by julek-wolfssl**
Fix a leak reported in https://github.com/wolfSSL/wolfssl/issues/4298

### CODE REVIEW

**Review by dgarske — APPROVED** [review]

**Comment by ordex** [comment]
I tested this branch and can confirm that the leak generated in wolfSSL_BIO_new() is now gone. Thanks!

**Comment by julek-wolfssl** [comment]
@dgarske Added a `atexit` call to the init. What do you think of this?

File: src/ssl.c [hunk 1]
```diff
@@ -4992,6 +4992,16 @@ int AddCA(WOLFSSL_CERT_MANAGER* cm, DerBuffer** pDer, int type, int verify)
 static int wolfSSL_RAND_InitMutex(void);
 #endif
 
+#ifdef OPENSSL_EXTRA
+static void AtExitCleanup(void)
+{
+    int i = initRefCount;
+    /* Call initRefCount times to make sure we perform cleanup */
+    for (; i > 0; i--)
+        (void)wolfSSL_Cleanup();
```
Comment (dgarske) [inline src/ssl.c:5001]: This should only be called by the calling application, not internally.

File: src/ssl.c [hunk 2]
```diff
@@ -5012,6 +5022,12 @@ int wolfSSL_Init(void)
 #endif
 
 #ifdef OPENSSL_EXTRA
+        /* OpenSSL registers cleanup using atexit */
+        if (atexit(AtExitCleanup) != 0) {
```
Comment (dgarske) [inline src/ssl.c:5026]: This is not portable and I would recommend against it.

**Review by dgarske — CHANGES_REQUESTED** [review]

File: src/ssl.c (hunk 2 above)
Comment (julek-wolfssl) [reply src/ssl.c:5026]: Added a configure.ac check for `atexit`. Would that work?

**Comment by julek-wolfssl** [comment]
`mem_buf` is also used with the pair BIO. Added to PR.

**Review by dgarske — APPROVED** [review]
I am okay with these changes. I would like another engineer like @ejohnstown also review before merge.

**Comment by ordex** [comment]
Just for the records, with this branch I cannot reproduce any of the reported memory leaks:

> ==24353== All heap blocks were freed -- no leaks are possible

**Comment by ejohnstown** [comment]
It is up to the user of the library to free the allocated memory, i.e., call `wolfSSL_Cleanup()`.

**Review by ejohnstown — CHANGES_REQUESTED** [review]
It is up to the user of the library to free the allocated memory, i.e., call wolfSSL_Cleanup().

**Comment by julek-wolfssl** [comment]
OpenSSL registers the cleanup call at initialization. Their documentation discourages calling the cleanup API so it is reasonable to expect that many projects using OpenSSL won't be calling it. `atexit` is only called on `OPENSSL_EXTRA && HAVE_ATEXIT` so it won't occur when someone is not using the compatibility layer.
 
About `OPENSSL_cleanup()` from https://www.openssl.org/docs/man1.1.1/man3/OPENSSL_cleanup.html
>Developers of libraries using OpenSSL are discouraged from calling this function and should instead, typically, rely on auto-deinitialisation.

File: src/ssl.c [hunk 3]
```diff
@@ -5004,6 +5004,16 @@ int AddCA(WOLFSSL_CERT_MANAGER* cm, DerBuffer** pDer, int type, int verify)
 static int wolfSSL_RAND_InitMutex(void);
 #endif
 
+#if defined(OPENSSL_EXTRA) && defined(HAVE_ATEXIT)
+static void AtExitCleanup(void)
+{
+    int i = initRefCount;
```
Comment (ejohnstown) [inline src/ssl.c:5010]: Why not set initRefCount to 1 and call wolfSSL_Cleanup once?

**Review by ejohnstown — CHANGES_REQUESTED** [review]
We are already using the constructor attribute in FIPS builds to force the self-test to run. Can you change this to use a function with the destructor attribute? It'll run automatically after exit, as well.

File: src/ssl.c (hunk 3 above)
Comment (julek-wolfssl) [reply src/ssl.c:5010]: Updated to your suggestion.

**Comment by julek-wolfssl** [comment]
Changed to use the gcc `destructor` attribute.

**Comment by julek-wolfssl** [comment]
Actually changed back to `atexit`. This makes supporting Windows and any other OS that supports `atexit` easy. If it doesn't have `atexit` then it needs to call the cleanup manually for OpenSSL as well anyway since it doesn't use compiler attributes for cleanup.

**Comment by julek-wolfssl** [comment]
Using the destructor attribute would require figuring out equivalents for multiple compilers. `atexit` is a system call that will work as long as it is detected by the configure script.

File: src/ssl.c (hunk 3 above)
Comment (dgarske) [reply src/ssl.c:5010]: Should the `if (initRefCount == 0) {` check be `> 0`? Guessing this code should only be called if the initRefCount is greater than zero.

**Review by dgarske — CHANGES_REQUESTED** [review]

File: src/ssl.c (hunk 3 above)
Comment (julek-wolfssl) [reply src/ssl.c:5010]: The cleanup routine will never decrement it to lower than 0 but I suppose this would bug proof this a little.

**Review by dgarske — APPROVED** [review]

**Comment by julek-wolfssl** [comment]
Retest this please.

**Review by ejohnstown — APPROVED** [review]

**Review by dgarske — APPROVED** [review]

**Pull request merged**